lib/Models/WebMapServiceCatalogItem.js

```diff
@@ -1099,15 +1101,51 @@ WebMapServiceCatalogItem.prototype._createImageryProvider = function(time) {
         }
     }
 
-    return new WebMapServiceImageryProvider({
+    const imageryOptions = {
         url : cleanAndProxyUrl(this, this.url),
         layers : this.layers,
         getFeatureInfoFormats : this.getFeatureInfoFormats,
         parameters : parameters,
         getFeatureInfoParameters : parameters,
         tilingScheme : defined(this.tilingScheme) ? this.tilingScheme : new WebMercatorTilingScheme(),
         maximumLevel: maximumLevel
-    });
+    };
+
+    var imageryProvider = new WebMapServiceImageryProvider(imageryOptions);
+
+    if (defined(this._lastToken)) {
+        // Using the last token is an optimization; if its still valid it will speed up
+        // the operation and if its not then it will just be requested when its needed.
+        imageryOptions.token = this._lastToken;
+    }
+
+    var maximumLevelBeforeMessage = maximumScaleToLevel(this.maximumScaleBeforeMessage);
+    if (defined(maximumLevelBeforeMessage)) {
+        var realRequestImage = imageryProvider.requestImage;
+        var messageDisplayed = false;
+
+        var that = this;
+        imageryProvider.requestImage = function(x, y, level) {
+            if (level > maximumLevelBeforeMessage) {
```

Comment:
Just to make sure I've understood this correctly, you're suggesting that we shouldn't use `maximumLevelBeforeMessage` and `showTilesAfterMessage` because they're more appropriate to be used in ArcGIS. Instead we can just use `hideLayerAfterMinScaleDenominator` and `maximumLevel` such that `if (level > maximumLevel)` we would only request `maximumLevel +1` tiles. Does that sound right?

Reply:
Yeah, I think you have that right. `maximumLevel` should be computed as it is currently, but if ``hideLayerAfterMinScaleDenominator` is true then the `WebMapServiceImageryProvider` should be constructed with `maxLevel+1` instead of `maxLevel`, and the `requestImage` override should be installed (as it is in your current implementation) to watch for requests for tiles with a level greater than maxLevel and display the message.